added architecture overview and guidance section (#39)
Azure-Samples/communication-services-authentication-hero-csharp: docs/design-guides/architecture-overview.md

> # Architecture Overview

Reviewer: How do I get to this page from the top-level README currently? The only way I found now is:
* Click on "Deploy to Azure using instructions as here"
* On that page scroll to the very bottom and click on "Proceed to Architecture Overview"

Is there a more direct way?

> # Architecture Overview

Author: We need to add that to READMe. Currently README of the repo needs to restructured , so that should go in there too.

# Architecture Design

> ## Table of content

Reviewer: nip: Table of content**s**

- [Overview](#overview)
- [Components](#components)
  - [Motivation for leveraging Graph Open Extensions for Identity Mapping](#motivation-for-leveraging-graph-open-extensions-for-identity-mapping)
- [Limitations](#limitations)
- [Guidance](#guidance)
- [Other Helpful Links to Explore](#other-helpful-links-to-explore)

## Overview

> This sample solution is primarily focused on building the ACS Authentication Backend Server backed by Azure Active Directory as the Identity Provider and using open source libraries from [Microsoft Identity Platform](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-overview). The sample solution can be used directly in below scenarios, otherwise the sample needs to be adapted as described in Guidance section below:

Reviewer 2: ```suggestion
This sample is primarily focused on building a Trusted Service for Azure Communication Services authentication. It is backed by Azure Active Directory as the Identity Provider and using open source libraries from [Microsoft Identity Platform](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-overview). The sample can be used directly in below scenarios, otherwise the sample needs to be adapted as described in Guidance section below:
```

> - The Azure Active Directory is configured to support single tenant use case. 

Reviewer: To me, the doc doesn't explain what exactly is the issue with the multi-tenant use-case...

> Note: You can verify the configuration through app registration used for user sign in flow for Azure Active Directory instance. Go to the specific app registration within Azure Active Directory through [Azure Portal](https://portal.azure.com/) and check the Authentication tab to verify the tenancy configuration.
- There are different Active Directory instances configured for different environments. 
> Note: The sample only supports identity mapping for single environment within Graph Open Extensions used for storage of identity mapping.

## Components
![Diagram](../images/ACS-Authentication-Server-Sample_Overview-Flow.png)

> As seen from the Overview Diagram, the key components of the solution are:

Reviewer: ```suggestion
As seen from the overview diagram, the key components of the sample are:
```

1. [Secure Web Api backed by Azure Active Directory](./secured-web-api-design.md)
2. [Identity Mapping leveraging Graph Open Extensions](./Identity-Mapping-Design_Graph-Open-Extensions.md)

> 3. Azure Communication Identity service which generates ACS Identity and Tokens. The interaction with ACS Identity service happens within [ACS sdk](https://docs.microsoft.com/en-us/azure/communication-services/concepts/sdk-options#sdks) used within the solution. 

Reviewer 2: ```suggestion
3. Azure Communication Identity service which generates Identity and Access Tokens. The sample uses the Azure Communication Services [Identity SDK](https://docs.microsoft.com/en-us/azure/communication-services/concepts/sdk-options#sdks).
```

> > Note: The `api/token/teams` endpoint does not leverage #2, as the M365 Azure Active Directory Identity is mapped to user's Teams Identity within ACS Identity Service.

Reviewer: ```suggestion
> Note: The `api/token/teams` endpoint does not leverage #2, as the M365 Azure Active Directory Identity is mapped to user's Teams Identity within Azure Communication Services, see [Custom Teams Endpoint documentation]().
```

### Motivation for leveraging Graph Open Extensions for Identity Mapping

> The primary motivation comes from the fact that the Azure Communication Identity for the user could be co-located with rest of the information for the Active Active Directory user. And as that would also prevent the usage of additional storage for storing ACS Identity mappings corresponding to Azure Active Directory user Id.

Reviewer 2: ```suggestion
The Azure Communication Services Identity for the user could be co-located with the information for the Active Active Directory user. This optimizes the complexity to maintain additional storage to keep mappings and instead enables developers to keep everything inside of Azure AD.
```

Reviewer: nit: this also has "Active Active Directory" 😄

## Limitations

> - The open extensions on Azure Active Directory can only be supported upto 2 per user instance as suggested [here](https://docs.microsoft.com/en-us/graph/extensibility-overview#open-extension-limits). Hence supporting multi tenancy could be a challenge as different extensions might be desired for different tenants.

Reviewer: ```suggestion
An application can add [at most two open extensions](https://docs.microsoft.com/en-us/graph/extensibility-overview#open-extension-limits) for an Azure Active Directory user.
```

> - There is Graph open extensions rate limit of 455 requests per 10 seconds as suggested [here](https://docs.microsoft.com/en-us/graph/throttling#open-and-schema-extensions-service-limits)

Reviewer: ```suggestion
Open extensions have a [rate limit](https://docs.microsoft.com/en-us/graph/throttling#open-and-schema-extensions-service-limits) of 455 requests per 10 seconds.
```

> - The sample solution leveraging Graph open extensions might not be suitable for the scenarios where the Azure Active Directory instance is used for 3rd party application sign in as with [delegated permissions granted over Graph Api](https://docs.microsoft.com/en-us/graph/auth/auth-concepts#delegated-and-application-permissions). The 3rd Party application in this case with right delegated permissions would be able to access the open extension data of the user.

Reviewer: This is unclear to me 😕

Author: This was something as an additional point from PM of the Open Extensions. So this is to say that if a 3rd party application is granted permissions on active directory instance of Company Contoso through app registration which has access to e.g. `user.read` permissions of contoso's   Active Directory user, then the open extensions data such as ACS user Id would be visible to 3rd party as well.
Since in this case ACS identity is EUPI, then Contoso should not be sharing that info. However I need to confirm this with Haider once he is back to office tomorrow/this week.

In general within team we feel it should not be concern given the active directory has more sensitive info e.g. address and stuff like that. So the sensitivity level of the ACS user Id should set the right guidance here for the users. Currently  its vague and left to user's choice how they want to handle that if they are fine sharing it or not.

> > Note: From data privacy perspective, the ACS Identity is categorized as [EUPI data](https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/properties/devenv-dataclassification-property#property-value)

Reviewer: Does this need to be here?

Author: From Architecture guidance perspective we need to let users know how they might want to make sure all the privacy compliance is met. May be there be Compliance sections in the doc.

Author: Actually this does not need to be here as already added on data privacy in guidance section. I will remove this part.

## Guidance

> 1. To support the multi environment scenario within Graph Open Extensions using the same Azure Active Directory user instance, you could extend the [IdentityMappingModel class](https://github.com/Azure-Samples/communication-services-authentication-hero-csharp/blob/main/src/Models/IdentityMappingModel.cs) to handle multiple environments.

Reviewer: This is also a bit unclear. Are we talking about storing mappings to multiple different `communicationUserIds`?

nit: "Open Extensions" vs "open extensions" capitalization is inconsistent throughout the doc.

Author: yeah this is the case of multiple ACS resources. However I have rephrased the documentation to focus only on our 1-1 identity mapping model. There is guidance provided to adjust the sample to different need.

> 2. If Graph open extensions does not meet your requirement for storage of Identity Mappings, you can consider storing them in databases as [CosmoDB](https://docs.microsoft.com/en-us/azure/cosmos-db/) or [Azure Tables](https://docs.microsoft.com/en-us/azure/storage/tables/) for an example. And you would need to update the write and read operations in [GraphService.cs class](https://github.com/Azure-Samples/communication-services-authentication-hero-csharp/blob/main/src/Services/GraphService.cs)

Reviewer: This I think we need to re-word a little bit. This whole section talks about alternative approach to storing the mappings in some kind of a database. I feel like instead of a bullet point this deserves a separate section "Other Identify Mapping Approaches".

nits:
* "CosmosDB",
* "ACS" -> "Azure Communication Services"

    - Replace the usage of Graph `/me/extensions` endpoint with [`/me`](https://docs.microsoft.com/en-us/graph/api/resources/users?view=graph-rest-1.0) endpoint. The `/me` endpoint will get the M365 user id in response as `id` attribute. This could be used as key to map the ACS Identity. The permissions for Graph Api on server app registrations just needs to be reduced to `User.Read`.
    - Persist the mapping of ACS Identity to M365 user id within the database of your choice.
    - Since the ACS Identity is classified as EUPI, please make sure all the required data privacy compliance are met on your end while the data is at rest and in transit.

> 3. Since Azure Communication Services is a data processor and you are the controller of the user data, you are responsible for ensuring the data privacy complaince. Please visit [here](https://docs.microsoft.com/en-us/azure/communication-services/concepts/privacy) for more information.

Reviewer: This feels like us transferring our internal terminology/complexity/frame of reference onto the user.

Author: Regarding compliance , I believe all companies should already be aware of how they need to make the data compliant , we just need to make sure to provide necessary handles to them. As in our sample case the Delete user endpoint is something to help our customers meet compliance. This was something Haider had provided feedback sometime back. 
Nevertheless I think this point may still need to rephrased well and may be needs more clarifications for users to make them actionable. I again plan to pick this up when Haider is back this week.

@ddematheu2 would you have additional thoughts here.

# Architecture Design

## Table of contents

- [Overview](#overview)
- [Components](#components)
  - [Motivation for leveraging Graph Open Extensions for Identity Mapping](#motivation-for-leveraging-graph-open-extensions-for-identity-mapping)
- [Limitations](#limitations)
- [Guidance](#guidance)
- [Alternate Identity Mapping Approach](#alternate-identity-mapping-approach)
- [Other Helpful Links to Explore](#other-helpful-links-to-explore)

## Overview

> This sample is primarily focused on building a Trusted Service for Azure Communication Services authentication. It is the ACS Authentication Backend Server backed by Azure Active Directory as the Identity Provider and using open source libraries from [Microsoft Identity Platform](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-overview). The sample can be used directly if the below conditions are met, otherwise the sample needs to be adapted as described in Guidance section below:

Reviewer 3: Remove locale from links

- The sample supports single tenant use case. 
> Note: You can verify the configuration through app registration used for user sign in flow for Azure Active Directory instance. Go to the specific app registration within Azure Active Directory through [Azure Portal](https://portal.azure.com/) and check the Authentication tab to verify the tenancy configuration.
- The sample only supports 1:1 identity mapping between Azure Active Directory user Id and Azure Communication Services user Id.

## Components
![Diagram](../images/ACS-Authentication-Server-Sample_Overview-Flow.png)

As seen from the overview diagram, the key components of the sample are:
1. [Secure Web Api backed by Azure Active Directory](./secured-web-api-design.md)
2. [Identity Mapping leveraging Graph Open Extensions](./Identity-Mapping-Design_Graph-Open-Extensions.md)

> 3. Azure Communication Identity service which generates Azure Communication Services Identity and Access Tokens. The sample uses the Azure Communication Services [Identity sdk](https://docs.microsoft.com/en-us/azure/communication-services/concepts/sdk-options#sdks). 

Reviewer 3: Remove locale from links

> > Note: The `api/token/teams` endpoint does not leverage #2, as the M365 Azure Active Directory Identity is internally mapped to user's Teams Identity within Azure Communication Services, see [Custom Teams Endpoint documentation](https://docs.microsoft.com/en-us/azure/communication-services/concepts/teams-endpoint).

Reviewer 3: Remove locale from links

### Motivation for leveraging Graph Open Extensions for Identity Mapping
The Azure Communication Services Identity for the user could be co-located with the information for the Azure Active Directory user. This optimizes the complexity to maintain additional storage to keep mappings and instead enables developers to keep everything inside of Azure AD.

## Limitations

> - An application can add [at most two open extensions](https://docs.microsoft.com/en-us/graph/extensibility-overview#open-extension-limits) for an Azure Active Directory user. 

Reviewer 3: Remove locale from links

> - Graph Open Extensions have a [rate limit](https://docs.microsoft.com/en-us/graph/throttling#open-and-schema-extensions-service-limits) of 455 requests per 10 seconds. 

Reviewer 3: Remove locale from links

## Guidance
1. If 1:1 identity model of Azure Active Directory and Azure Communication Services in the sample does not meet your requirement, then you can consider adapting the [IdentityMappingModel class](https://github.com/Azure-Samples/communication-services-authentication-hero-csharp/blob/main/src/Models/IdentityMappingModel.cs) to handle multiple identity mappings.
2. Since Azure Communication Services is a data processor and you are the controller of the user data, you are responsible for ensuring the data privacy compliance. Please visit [here](https://docs.microsoft.com/en-us/azure/communication-services/concepts/privacy) for more information.
3. **For information of the users:** When the Azure Active Directory instance is used for 3rd party application sign in with [delegated permissions granted over Graph Api](https://docs.microsoft.com/en-us/graph/auth/auth-concepts#delegated-and-application-permissions), the 3rd Party application with delegated permissions as `user.read` would also have access to the Azure Communication Services user Id persisted as open extension data of the user.

## Alternate Identity Mapping Approach

> If Graph Open Extensions does not meet your requirement for storage of Identity Mappings, you can consider storing them in databases as [CosmosDB](https://docs.microsoft.com/en-us/azure/cosmos-db/) or [Azure Tables](https://docs.microsoft.com/en-us/azure/storage/tables/) for an example. You will need to make below changes:

Reviewer 3: Remove locale from links

1. Update the write and read operations in [GraphService.cs class](https://github.com/Azure-Samples/communication-services-authentication-hero-csharp/blob/main/src/Services/GraphService.cs)

> 2. Replace the usage of Graph `/me/extensions` endpoint with [`/me`](https://docs.microsoft.com/en-us/graph/api/resources/users?view=graph-rest-1.0) endpoint. The `/me` endpoint will get the M365 user Id in response as `id` attribute. This could be used as key to map the ACS Identity. The permissions for Graph Api on server app registrations just needs to be reduced to `User.Read`.

Reviewer 3: Remove locale from links

3. Persist the mapping of ACS Identity to M365 user Id within the database of your choice.
4. Since the Azure Communication Services Identity is classified as EUPI, please make sure all the required data privacy compliance are met on your end while the data is at rest and in transit.

## Other Helpful Links to Explore
- [Graph Apis](https://docs.microsoft.com/en-us/graph/use-the-api)
- [Example on how M365 handles data rentention policy](https://docs.microsoft.com/en-us/compliance/assurance/assurance-data-retention-deletion-and-destruction-overview#data-retention) to meet the Data Privacy requirements for different data category.
- [Microsoft Identity Platform](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-overview)

> - [Azure Communication Services Documentation](https://docs.microsoft.com/en-us/azure/communication-services/)

Reviewer 3: Remove locale from links
